src/operator/subgraph/dnnl/dnnl_fc.cc

```diff
@@ -72,83 +75,123 @@ class SgDNNLFCOp {
-  size_t base_num_inputs  = has_bias ? 3 : 2;
-  size_t base_num_outputs = 1;
-
-  float min_data   = 0.0f;
-  float max_data   = 0.0f;
-  float min_weight = 0.0f;
-  float max_weight = 0.0f;
-  float min_bias   = 0.0f;
-  float max_bias   = 0.0f;
-
-  if (!initialized_) {
-    if (dnnl_param.channel_wise_quantize.has_value() && dnnl_param.channel_wise_quantize) {
-      channel_wise_runtime_ = true;
+  auto& dnnl_param         = full_param_.dnnl_param;
+  auto& default_param      = full_param_.default_param;
+  const bool has_bias      = !default_param.no_bias;
+  const bool quantized     = dnnl_param.quantized;
+  const bool out_quantized = dnnl_param.quantized && !dnnl_param.enable_float_output;
+  const bool channel_wise  = quantized && dnnl_param.channel_wise_quantize.has_value() &&
+                            dnnl_param.channel_wise_quantize.value();
+
+  const FCInputIndex idx(full_param_);
+
+  CHECK_EQ(in_data.size(), idx.GetTotal());
+
+  int index               = 0;
+  const int out_index     = index++;
+  const int out_min_index = out_quantized ? index++ : 0;
+  const int out_max_index = out_quantized ? index++ : 0;
+  CHECK_EQ(out_data.size(), index);  // index is equal to total number of outputs
+
+  float data_min   = 0.0f;
+  float data_max   = 0.0f;
+  float weight_min = 0.0f;
+  float weight_max = 0.0f;
+  float bias_min   = 0.0f;
+  float bias_max   = 0.0f;
+
+  const float sum_min   = idx.sum_min ? in_data[idx.sum_min].data().dptr<float>()[0] : 0.0;
+  const float sum_max   = idx.sum_max ? in_data[idx.sum_max].data().dptr<float>()[0] : 0.0;
+  NDArray data          = in_data[idx.data];
+  const NDArray& weight = in_data[idx.weight];
+  NDArray output;
+
+  if (dnnl_param.with_sum) {
+    if (!initialized_) {
+      // TODO(zhennan): Currently, dnnl fallback mechanism will break inplace option,
+      // which make check (req[out_index] == kWriteInplace) useless.
+      auto in_dnnl_mem  = static_cast<const dnnl::memory*>(in_data[idx.sum].GetDNNLData());
+      auto out_dnnl_mem = static_cast<const dnnl::memory*>(out_data[out_index].GetDNNLData());
+      if (in_dnnl_mem->get_data_handle() == out_dnnl_mem->get_data_handle()) {
+        inplace_ = true;
+      }
     }
-
-    total_num_inputs_  = base_num_inputs;
-    total_num_outputs_ = base_num_outputs;
-    if (dnnl_param.quantized) {
-      total_num_inputs_ = channel_wise_runtime_ ? (base_num_inputs + 2) : (base_num_inputs * 3);
-      total_num_outputs_ =
-          dnnl_param.enable_float_output ? base_num_outputs : (base_num_outputs * 3);
+    if (inplace_) {
+      output = in_data[idx.sum];
+    } else {
+      // Not in place: copy in_data[idx.sum] into outputs[out_index].
+      auto in_dnnl_mem  = static_cast<const dnnl::memory*>(in_data[idx.sum].GetDNNLData());
+      auto out_dnnl_mem = static_cast<const dnnl::memory*>(out_data[out_index].GetDNNLData());
+      if (out_data[out_index].dtype() == mshadow::kInt32) {
```

Comment:
Is it taken under consideration what if the output of the fc is a float and the data to sum with is an integer? Shouldn't there also be a check for mshadow::Float32 or isn't it a proper scenario?

Reply:
For now such configuration doesn't exist. For 'full' quantization mode input and output are int8 and for 'smart' quantization both are float